Abl duck

spazzymoto opened the pull request:
Finally finished, sorry for the delay. I wouldn't say Java is my strongest language so any feedback would be appreciated. Tried to remove as much duplication as i can by adding the getClassCompilationUnit method to your ClassDocumentationVisitor class.

spazzymoto committed:
Initial commit

First commit of code refactored for PCT

spazzymoto committed:
Improved javadoc syntax error messages

Included the source class file, method or property whose comment generated the syntax error.

spazzymoto committed:
Method override bug

When you have multiple methods with the same names it cause the html to generate multiple elements with the same id, which is not allowed. This fixes that

spazzymoto committed:
Remove redundant javadoc.jar

Moved the class to the za.co.mip.ablduck package

spazzymoto committed:
Fix indentation errors

gquerret commented:
Thanks !
* First comment after a quick look (in the GitHub UI): you should include the .g4 so that Java classes can be generated from the PCT build script.
* And another one, on the resources, you included a ZIP file, it would be better to store directly the content rather than the ZIP itself
* Any unit tests ?

spazzymoto commented:
You are correct the better way would be to add the g4 files. 

I was originally going to have another repo with the template in it and was going to minify the files before including in the build. But I can investigate doing this via ant in the PCT build. 

No unit tests yet :|, will get on that asap

Thanks for the feedback :)

spazzymoto committed:
Generate javadoc classes 

Generate javadoc classes from the antlr grammar before compile.

spazzymoto committed:
Added Template directory

Added template to /za/co/mip/ablduck/template

spazzymoto committed:
Fix SonarQube results

Various fixeds for sonarqube results.
Changed javadoc to generate each build

spazzymoto committed:
Sonarqube errors

spazzymoto committed:
Renamed listener property to errorListener

Had a variable named listener already.

spazzymoto committed:
Fixed minor sonarqube errors

spazzymoto committed:
Remove unneeded java.lang import

spazzymoto committed:
Data files only

Add parameter to allow the generation of data files only

spazzymoto committed:
Add Unit Test

Added some basic unit tests

spazzymoto committed:
Does this fix my IllegalAccessException

spazzymoto committed:
Try again

spazzymoto committed:
Remove system.out.println

gquerret commented:
Some more comments:
* IIRC, what you've done is based on documentation for another language, what is the licence of all files under src/java/za/co/mip/ablduck/templates/app ? Wouldn't it be better to download them from a CDN at runtime ?
* ExtJS and prettify are probably better downloaded from a CDN too. ExtJS is GPL v3 which may conflict with the Apache licence of the PCT project. openedge-language seems to be an extension of prettify, what's the licence ?

spazzymoto committed:
Remove GPL code

spazzymoto committed:
SonarQube Errors

spazzymoto committed:
Improve HTML generation

spazzymoto committed:
Remove tabs in properties file

spazzymoto committed:
Provide setters/getters for CommentParser

spazzymoto committed:
Fix null pointer exeception on parameter comments

gquerret commented:
I'll ask for a few more changes (if you're ready to merge on your side):
* include antlr-runtime in PCT.jar as it's much smaller than antlr-complete
* format the source code in Eclipse (Ctrl Maj F, that should be preconfigured for PCT conventions)

gquerret commented on build.xml:
```diff
@@ -80,6 +81,11 @@
+    <!-- ABLDuck classes -->
+    <pathelement location="lib/antlr-4.7-complete.jar"/>
+    <pathelement location="lib/markdown4j-2.2.jar"/>
+  	<pathelement location="${build-javadoc}" />
```
Don't you know you'd have to use spaces, not tabs ?
https://arstechnica.com/information-technology/2017/06/according-to-statistics-programming-with-spaces-instead-of-tabs-makes-you-rich/

spazzymoto replied:
Haha ah! Thats where i've been going wrong :rofl: sorry i normally have my editor set to replace tabs with spaces but i see i didn't do this after reinstalling.

gquerret commented on .classpath:
```diff
@@ -13,6 +13,7 @@
 	<classpathentry kind="lib" path="lib/ast-dependencies.jar"/>
 	<classpathentry kind="lib" path="lib/ini4j-0.5.1.jar"/>
+	<classpathentry kind="lib" path="lib/antlr-4.7-complete.jar"/>
+	<classpathentry kind="lib" path="lib/markdown4j-2.2.jar"/>
```
What is markdown used for ?

spazzymoto replied:
Markdown is the library that converts the markdown in your comment to HTML

gquerret commented on src/java/eu/rssw/pct/oedoc/OpenEdgeDocumentation115.java:
```diff
@@ -118,7 +120,7 @@ public void execute() {
         // Destination directory must exist
         if (this.destDir == null) {
-            throw new BuildException(Messages.getString("OpenEdgeClassDocumentation.0"));
+            throw new BuildException(MessageFormat.format(Messages.getString("OpenEdgeClassDocumentation.0"), "destDir"));
```
Just to be sure, do you re-use code from this task ?

spazzymoto replied:
I currently took a copy of the OpenEdgeDocumentation115.java class to write ABLDuck.java, i could look at inheriting it instead :+1:

Commit:
Classpath

spazzymoto committed:
Fix searching

Also formatted the source code and only including the antlr runtime in the PCT.jar

spazzymoto committed:
Merge remote-tracking branch 'origin/ABLDuck' into ABLDuck

spazzymoto commented:
I would like to get the interfaces and events in before you merge if thats ok? Just to have a more complete solution

spazzymoto committed:
New markdown parser & Support for old style comments

Switched the markdown processor to http://spec.commonmark.org for better markdown support. Also added the experimental ability to detect old style comments and parse them. Each tag: text witll be turned into ## tag \n text to be rendered nicely by the markdown processor.

spazzymoto committed:
Various bug fixes

spazzymoto committed:
Added the implement & implementers sections

Classes will show you what interfaces they implement and interfaces will show you which classes implement them.
Also added a hack to determine the fully qualified class name from a using statement blegh

spazzymoto committed:
Changed logo and favicon

Plus random sonar rules violations

spazzymoto committed:
Fix author

spazzymoto commented:
If you happy to merge so am i, will wait for feedback before i spend anymore time :+1:

gquerret committed:
Exception
